### ISSUE TITLE
Error: Invalid URI

### ISSUE THREAD

**Issue by castorpollux285** [donavanbecker/homebridge-meross#20]
Hello,

I get the following error message when I try to configure the meross plug "mss310":

```javascript
[4/12/2020, 11:25:27 AM] [Meross_device] Failed to POST to the Meross Plug: Error: Invalid URI "192.168.178.35/config"
    at Request.init (/usr/local/lib/node_modules/homebridge-meross/node_modules/request/request.js:273:31)
    at new Request (/usr/local/lib/node_modules/homebridge-meross/node_modules/request/request.js:127:8)
    at request (/usr/local/lib/node_modules/homebridge-meross/node_modules/request/index.js:53:10)
    at /usr/local/lib/node_modules/homebridge-meross/index.js:9:5
    at new Promise (<anonymous>)
    at doRequest (/usr/local/lib/node_modules/homebridge-meross/index.js:8:10)
    at MerossPlug.getOnCharacteristicHandler (/usr/local/lib/node_modules/homebridge-meross/index.js:175:24)
    at Characteristic.On.emit (events.js:310:20)
    at Characteristic.On.Characteristic.getValue (/usr/local/lib/node_modules/homebridge/node_modules/hap-nodejs/lib/Characteristic.js:164:10)
    at Bridge.<anonymous> (/usr/local/lib/node_modules/homebridge/node_modules/hap-nodejs/lib/Accessory.js:771:20)
```

This is the configuration:
```json
{
            "name": "Meross_device",
            "deviceUrl": "192.168.178.35",
            "hardwareVersion": 2,
            "channel": 0,
            "messageId": "242afa8372dd827e10ee53de77fdfc35",
            "timestamp": 1586678958,
            "sign": "40f1117a6bf82cae9757915ff8a9a564",
            "accessory": "Meross"
        }
```

I got the informations from the Charles ios app.

Thanks in advance!

**Comment by donavanbecker** [comment]
@castorpollux285 are you using config-ui-x?

**Comment by arohl** [comment]
@donavanbecker Is it possible to reopen this? I have the meross plug "mss310" and get a similar error (although I got my parameters using https://github.com/bapirex/meross-api/). I did use config-ui-x.

Many thanks

  Andrew

[4/16/2020, 10:07:09 PM] [Meross] Failed to POST to the Meross Plug: Error: Invalid URI "undefined/config"
    at Request.init (/usr/local/lib/node_modules/homebridge-meross/node_modules/request/request.js:273:31)
    at new Request (/usr/local/lib/node_modules/homebridge-meross/node_modules/request/request.js:127:8)
    at request (/usr/local/lib/node_modules/homebridge-meross/node_modules/request/index.js:53:10)
    at /usr/local/lib/node_modules/homebridge-meross/index.js:9:5
    at new Promise (<anonymous>)
    at doRequest (/usr/local/lib/node_modules/homebridge-meross/index.js:8:10)
    at MerossPlug.getOnCharacteristicHandler (/usr/local/lib/node_modules/homebridge-meross/index.js:175:24)
    at Characteristic.On.emit (events.js:310:20)
    at Characteristic.On.Characteristic.getValue (/usr/local/lib/node_modules/homebridge/node_modules/hap-nodejs/lib/Characteristic.js:164:10)
    at Bridge.<anonymous> (/usr/local/lib/node_modules/homebridge/node_modules/hap-nodejs/lib/Accessory.js:771:20)
[4/16/2020, 10:07:09 PM] [Meross] Retrieved status unsuccessfully.
[4/16/2020, 10:07:09 PM] [Meross] calling getOnCharacteristicHandler: false
[4/16/2020, 10:07:11 PM] [Meross] {
  name: 'Meross',
  accessories: [
    {
      name: 'Amp',
      deviceUrl: '192.168.178.139',
      hardwareVersion: 2,
      channel: 0,
      messageId: 'XXX',
      timestamp: XXX,
      sign: 'XXX'
    }
  ],
  accessory: 'Meross'
} undefined/config

**Comment by dylanfrankcom** [comment]
@castorpollux285 @arohl 

Could you both edit your deviceUrl in your config file to include the http:// header and see if that fixes the issue. 

Ex. `deviceUrl: 'http://192.168.178.139'`

**Comment by arohl** [comment]
Will do first thing tomorrow- it’s quite late in Perth :-)

However I did manage to get Charles on my iPhone and the POST seems very simple for this model of smart plug

{
    "payload": {
        "togglex": {
            "onoff": 1,
            "channel": 0
        }
    },
    "header": {
        "messageId": "164d94c5968c397bd17a90f1797f8650",
        "method": "SET",
        "from": "http:\/\/192.168.178.139\/config",
        "sign": "7028071d19f39430d61842b9cf303f94",
        "namespace": "Appliance.Control.ToggleX",
        "triggerSrc": "iOS",
        "timestamp": 1587046811,
        "payloadVersion": 1
    }
}

**Comment by dylanfrankcom** [comment]
Thanks for posting the payload. It will help later if the changes in your config don't. Please report back if it works!

**Comment by arohl** [comment]
Got the same error
`
[Meross] Failed to POST to the Meross Plug: Error: Invalid URI "undefined/config"
    at Request.init (/usr/local/lib/node_modules/homebridge-meross/node_modules/request/request.js:273:31)
    at new Request (/usr/local/lib/node_modules/homebridge-meross/node_modules/request/request.js:127:8)
    at request (/usr/local/lib/node_modules/homebridge-meross/node_modules/request/index.js:53:10)
    at /usr/local/lib/node_modules/homebridge-meross/index.js:9:5
    at new Promise (<anonymous>)
    at doRequest (/usr/local/lib/node_modules/homebridge-meross/index.js:8:10)
    at MerossPlug.getOnCharacteristicHandler (/usr/local/lib/node_modules/homebridge-meross/index.js:175:24)
    at Characteristic.On.emit (events.js:310:20)
    at Characteristic.On.Characteristic.getValue (/usr/local/lib/node_modules/homebridge/node_modules/hap-nodejs/lib/Characteristic.js:164:10)
    at Bridge.<anonymous> (/usr/local/lib/node_modules/homebridge/node_modules/hap-nodejs/lib/Accessory.js:771:20)
[4/17/2020, 12:32:14 PM] [Meross] Retrieved status unsuccessfully.
[4/17/2020, 12:32:14 PM] [Meross] calling getOnCharacteristicHandler: false
[4/17/2020, 12:32:16 PM] [Meross]
`
 {
  name: 'Meross',
  accessories: [
    {
      name: 'Amp',
      deviceUrl: 'http://192.168.178.139',
      hardwareVersion: 2,
      channel: 0,
      messageId: 'XXXXX',
      timestamp: 1587043076,
      sign: 'XXXXXX'
    }
  ],
  accessory: 'Meross'
} undefined/config

**Comment by dylanfrankcom** [comment]
I think the issue is with your config file. I believe JSON strings can only be declared with double quotes as opposed to single quotes and I think there is an issue with the layout in general. If you want you can post your entire config with a couple things redacted and I can look over it if you would like? The bridge's mac-address `username` value as well as any devices `messageID` and `sign` values can be removed or edited. If you don't feel comfortable posting feel free to message me on [twitter](https://twitter.com/dylanfrankcom) and we can figure something out. 🙂

**Comment by arohl** [comment]
I think its much harder than this unfortunately. I have been playing around with this plug and can now send it data but I always get an error.

{
 header: {
   messageId: 'c3222c7d2b9163fe2968f06c45338a9f',
   namespace: 'Appliance.Control.ToggleX',
   method: 'ERROR',
   payloadVersion: 1,
   from: '/appliance/19090313123344251h1448e1e900e715/publish',
   timestamp: 1587132239,
   timestampMs: 245,
   sign: '7c8b932061c9cea12444284cc7d2626c'
 },
 payload: { error: { code: 5001, detail: 'sign error' } }

When I look at what is being sent from the Meross App to the plug, the sign changes everytime so I am guessing that they have improved their security and until I find out how the sign is generated, I am out of luck :-(

**Comment by dylanfrankcom** [comment]
You’re right! The sign does change with every request sent through the Meross app, however a unique sign is not required with every new payload sent. The sign is able to be reused and is why it is included in the config file. I am not completely sure if the messageID and the sign are associated with each other and must be taken from the same payload to work correctly but is something else you can try. 

So far, this is expected behaviour and happens on all Meross devices. This also means there is still hope and I believe your device should be able to work.

**Comment by arohl** [comment]
OK

SO here is my config (copy and pasted from web gui)

{
    "bridge": {
        "name": "Homebridge A578",
        "username": "0E:35:C2:82:A5:78",
        "port": 51901,
        "pin": "XXXXXXXXX"
    },
    "accessories": [
        {
            "accessory": "CmdTriggerSwitch",
            "name": "amp",
            "stateful": true,
            "onCmd": "/Users/andrew/src/MSS310/start_amp.sh",
            "offCmd": "/Users/andrew/src/MSS310/stop_amp.sh"
        },
        {
            "name": "Meross",
            "accessories": [
                {
                    "name": "amp2",
                    "deviceUrl": "192.167.178.139",
                    "hardwareVersion": 2,
                    "channel": 0,
                    "messageId": "17cfbf76eb100a721715de2c617f87fd",
                    "timestamp": 1587218340,
                    "sign": "XXXXXXXXX"
                }
            ],
            "accessory": "Meross"
        }
    ],
    "platforms": [
        {
            "name": "Config",
            "port": 8581,
            "platform": "config"
        }
    ]
}

**Comment by arohl** [comment]
Note that I have got something working using CmdTriggerSwitch - it calls a shell script which in turn uses a python library I found BUT this does it via talking to the meross cloud so has a lot of latency

**Comment by arohl** [comment]
Errors still like:

[4/18/2020, 10:09:02 PM] [Meross] Failed to POST to the Meross Plug: Error: Invalid URI "undefined/config"
    at Request.init (/usr/local/lib/node_modules/homebridge-meross/node_modules/request/request.js:273:31)
    at new Request (/usr/local/lib/node_modules/homebridge-meross/node_modules/request/request.js:127:8)
    at request (/usr/local/lib/node_modules/homebridge-meross/node_modules/request/index.js:53:10)
    at /usr/local/lib/node_modules/homebridge-meross/index.js:9:5
    at new Promise (<anonymous>)
    at doRequest (/usr/local/lib/node_modules/homebridge-meross/index.js:8:10)
    at MerossPlug.getOnCharacteristicHandler (/usr/local/lib/node_modules/homebridge-meross/index.js:175:24)
    at Characteristic.On.emit (events.js:310:20)
    at Characteristic.On.Characteristic.getValue (/usr/local/lib/node_modules/homebridge/node_modules/hap-nodejs/lib/Characteristic.js:164:10)
    at Bridge.<anonymous> (/usr/local/lib/node_modules/homebridge/node_modules/hap-nodejs/lib/Accessory.js:771:20)

**Comment by donavanbecker** [comment]
@arohl Your device url still doesn’t have http:// in front of it.

**Comment by arohl** [comment]
OK - its all working now! - thanks!!

I hand edited the config to add the http and also to simplify it (it didn't work with the http added to my previous config); it looks like

    "accessories": [
        {
            "accessory": "Meross",
            "name": "amp",
            "deviceUrl": "http://192.168.178.139",
            "hardwareVersion": 2,
            "channel": 0,
            "messageId": "17cfbf76eb100a721715de2c617f87fd",
            "timestamp": 1587218340,
            "sign": "XXXXXXXXXX"
        }
    ],

**Comment by dylanfrankcom** [comment]
Fantastic! Was about to send you the exact same thing. Enjoy! 🙂

**Comment by donavanbecker** [comment]
@arohl would you be willing to test a change that I just made to the beta branch?

**Comment by arohl** [comment]
> @arohl would you be willing to test a change that I just made to the beta branch?

Sure - but you will have to give me some basic instructions on how to install beta branch

**Comment by donavanbecker** [comment]
thanks,

from terminal install beta using this command:
```
sudo npm install -g --unsafe-perm homebridge-meross@beta
```
Then in your config schema, change the Device URL to just be an IP address. without the http:// in front of it. and let me know if you are still able to control your device?

**Comment by arohl** [comment]
Can confirm it works!

Once you have updated the repo, I just need to

`sudo npm install -g --homebridge-meross`

to get back on the release version?

**Comment by donavanbecker** [comment]
with `@latest` at the end just to make sure it pulls the `latest` from npm.

**Comment by donavanbecker** [comment]
Thanks fro doing that!

**Comment by donavanbecker** [comment]
Just released [Version 1.1.0](https://github.com/donavanbecker/homebridge-meross/releases/tag/v1.1.0), if you want to give it a try.

**Comment by arohl** [comment]
Done and can confirm works for me

On 29 Apr 2020, at 4:50 am, Donavan Becker <notifications@github.com<mailto:notifications@github.com>> wrote:



Just released Version 1.1.0<https://github.com/donavanbecker/homebridge-meross/releases/tag/v1.1.0>, if you want to give it a try.
